x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/dataframe/transforms/pivot/DateHistogramGroupSource.java

```diff
@@ -248,6 +256,71 @@ public static DateHistogramGroupSource fromXContent(final XContentParser parser,
         return lenient ? LENIENT_PARSER.apply(parser, null) : STRICT_PARSER.apply(parser, null);
     }
 
+    public static boolean isInvalidFormat(String format, DateHistogramInterval interval) {
+        if (format == null || interval == null) {
+            return false;
+        }
+        DateFormatter formatter = DateFormatter.forPattern(format);
+        String expression = interval.toString();
+        String epochFormat = formatter.formatMillis(Instant.EPOCH.toEpochMilli());
+        String epochPlusIntervalFormat = formatter.formatMillis(Instant.EPOCH.toEpochMilli() + interval.estimateMillis());
+        if (DateHistogramAggregationBuilder.DATE_FIELD_UNITS.containsKey(expression)) {
+            // The estimated rounding defaults to 30 days for monthly, EPOCH starts in January, which has 31 days.
+            if (Rounding.DateTimeUnit.MONTH_OF_YEAR.equals(DateHistogramAggregationBuilder.DATE_FIELD_UNITS.get(expression))) {
+                epochPlusIntervalFormat = formatter.formatMillis(Instant.EPOCH.toEpochMilli() + THIRTY_ONE_DAYS_MS);
+            }
+        } else {
+            // This should never really happen in practice. date_histogram does not support multiple month intervals
+            if (expression.endsWith("M")) {
+                throw new IllegalArgumentException("only single month intervals are supported");
+            }
+            final String normalized = expression.toLowerCase(Locale.ROOT).trim();
+            long millisecondLength;
+            if (normalized.endsWith("ms")) {
+                millisecondLength = unitMilliseconds(normalized, "ms", TimeUnit.MILLISECONDS).getMillis();
+            } else if (normalized.endsWith("s")) {
```

Comment:
Continuing the idea above this bit becomes:

```
    Duration duration;
    if (normalized.endsWith("ms")) {
        duration = Duration.ofMillis(parse(normalized, "ms"));
    } else if (normalized.endsWith("s")) {
        duration = Duration.ofSeconds(parse(normalized, "s"));
    } else if (normalized.endsWith("m")) {
        // similar pattern for all the others too
    } else {
        return true;
    }
    epochPlusIntervalFormat = formatter.formatMillis(Instant.EPOCH.plus(duration));
```

Then you don't need the `unitMilliseconds()` method at all.